### ISSUE TITLE
Enabling cgroup metrics in kubernetes causes job failures

### ISSUE THREAD

**Issue by nuwang** [galaxyproject/galaxy#12366]
**Describe the bug**
Enabling cgroup metrics in `job_metrics_conf.xml` breaks jobs in k8s. Specifically the last line in the job script results in an error, for which the stdout is swallowed, and consequently, it causes the job  as a whole to fail.

There seem to be at least three issues here.
a. failing cgroup metrics should probably not swallow stderr?
b. cgroup metrics should work when running within a k8s container.
c. The k8s runner should raise a better error message when the main job script returns an error code. Specifically, it just fails with: ```More pods failed than allowed. See stdout for pods details.```. Since the tool itself has succeeded, there is no stderr.

**Galaxy Version and/or server at which you observed the bug**
Galaxy Version: 21.05

**To Reproduce**
1. Enable job_metrics_config_file in galaxy.yml
2. Add the following contents to it:
```
                job_metrics_conf.yml:
                  - type: core
                  - type: cgroup
```
3. Run job in k8s
4. Further investigation reveals that the line in question that triggers the issue is:
```
$ if [ -e "/proc/$$/cgroup" -a -d "/sys/fs/cgroup" ]; then     cgroup_path=$(cat "/proc/$$/cgroup" | awk -F':' '$2=="memory"{print $3}');     for f in /sys/fs/cgroup/memory$cgroup_path/memory.*; do         echo "__$(basename $f)__" >> /galaxy/server/database/jobs_directory/000/698/__instrument_cgroup__metrics; cat "$f" >> /galaxy/server/database/jobs_directory/000/698/__instrument_cgroup__metrics;     done; fi
```
The result is:
```
cat: '/sys/fs/cgroup/memory/kubepods/burstable/podd99448ee-b71c-4263-8f8c-d0845e8d115b/688ad50f38eede052508815a4a0cf964b0cc961eae97815fef9a99c21c331532/memory.*': No such file or directory
```

Specifically, it appears that the `$cgroup_path` identified by the script (`/sys/fs/cgroup/memory/kubepods/burstable/podd99448ee-b71c-4263-8f8c-d0845e8d115b/688ad50f38eede052508815a4a0cf964b0cc961eae97815fef9a99c21c331532/memory.`) does not exist within the container.
```
$ cat /proc/$$/cgroup
12:hugetlb:/kubepods/burstable/podd99448ee-b71c-4263-8f8c-d0845e8d115b/688ad50f38eede052508815a4a0cf964b0cc961eae97815fef9a99c21c331532
11:pids:/kubepods/burstable/podd99448ee-b71c-4263-8f8c-d0845e8d115b/688ad50f38eede052508815a4a0cf964b0cc961eae97815fef9a99c21c331532
10:cpuset:/kubepods/burstable/podd99448ee-b71c-4263-8f8c-d0845e8d115b/688ad50f38eede052508815a4a0cf964b0cc961eae97815fef9a99c21c331532
9:devices:/kubepods/burstable/podd99448ee-b71c-4263-8f8c-d0845e8d115b/688ad50f38eede052508815a4a0cf964b0cc961eae97815fef9a99c21c331532
8:freezer:/kubepods/burstable/podd99448ee-b71c-4263-8f8c-d0845e8d115b/688ad50f38eede052508815a4a0cf964b0cc961eae97815fef9a99c21c331532
7:memory:/kubepods/burstable/podd99448ee-b71c-4263-8f8c-d0845e8d115b/688ad50f38eede052508815a4a0cf964b0cc961eae97815fef9a99c21c331532
6:rdma:/
5:blkio:/kubepods/burstable/podd99448ee-b71c-4263-8f8c-d0845e8d115b/688ad50f38eede052508815a4a0cf964b0cc961eae97815fef9a99c21c331532
4:perf_event:/kubepods/burstable/podd99448ee-b71c-4263-8f8c-d0845e8d115b/688ad50f38eede052508815a4a0cf964b0cc961eae97815fef9a99c21c331532
3:cpu,cpuacct:/kubepods/burstable/podd99448ee-b71c-4263-8f8c-d0845e8d115b/688ad50f38eede052508815a4a0cf964b0cc961eae97815fef9a99c21c331532
2:net_cls,net_prio:/kubepods/burstable/podd99448ee-b71c-4263-8f8c-d0845e8d115b/688ad50f38eede052508815a4a0cf964b0cc961eae97815fef9a99c21c331532
1:name=systemd:/kubepods/burstable/podd99448ee-b71c-4263-8f8c-d0845e8d115b/688ad50f38eede052508815a4a0cf964b0cc961eae97815fef9a99c21c331532
0::/system.slice/rke2-server.service
```
Instead, the cgroup metrics appear to be living in the container under:

```
$ ls /sys/fs/cgroup/memory/
cgroup.clone_children  cgroup.procs    memory.force_empty   memory.kmem.limit_in_bytes      memory.kmem.slabinfo     memory.kmem.tcp.limit_in_bytes      memory.kmem.tcp.usage_in_bytes  memory.limit_in_bytes      memory.memsw.failcnt         memory.memsw.max_usage_in_bytes  memory.move_charge_at_immigrate  memory.oom_control     memory.soft_limit_in_bytes  memory.swappiness      memory.use_hierarchy  tasks
cgroup.event_control   memory.failcnt  memory.kmem.failcnt  memory.kmem.max_usage_in_bytes  memory.kmem.tcp.failcnt  memory.kmem.tcp.max_usage_in_bytes  memory.kmem.usage_in_bytes      memory.max_usage_in_bytes  memory.memsw.limit_in_bytes  memory.memsw.usage_in_bytes      memory.numa_stat                 memory.pressure_level  memory.stat                 memory.usage_in_bytes  notify_on_release
```

**Additional context**
1. Core metrics work fine

**Comment by natefoo** [comment]
For containers where the non-namespaced `/sys/fs/cgroup` is probably the container cgroup we should probably just override the path handling stuff. Can you verify that the values in e.g. `/sys/fs/cgroup/memory/memory.memsw.limit_in_bytes` or `memory.memsw.max_usage_in_bytes` are specific to that container and not the same as the values in those files on the host?

**Comment by nuwang** [comment]
@natefoo That seems to be the correct path (also supported by this SO post: https://stackoverflow.com/questions/42187085/check-mem-limit-within-a-docker-container). However, while the container memory limits are reported correctly, the paths are different, because swap is not supported in k8s yet (https://kubernetes.io/blog/2021/08/09/run-nodes-with-swap-alpha/): 

```
galaxy@galaxy-job-0-7ff64656dc-2z5jc:/galaxy/server$ cat /sys/fs/cgroup/memory/memory.limit_in_bytes
6999998464
galaxy@galaxy-job-0-7ff64656dc-2z5jc:/galaxy/server$ cat /sys/fs/cgroup/memory/`
9223372036854771712
galaxy@galaxy-job-0-7ff64656dc-2z5jc:/galaxy/server$ cat /sys/fs/cgroup/memory/memory.memsw.max_usage_in_bytes
424902656
galaxy@galaxy-job-0-7ff64656dc-2z5jc:/galaxy/server$ cat /sys/fs/cgroup/memory/memory.max_usage_in_bytes
424902656
```

So `memory.limit_in_bytes` is correct at 7Gi, which is what we set, but `memory.memsw.limit_in_bytes` is just an arbitrary limit by the looks of it. `memory.memsw.max_usage_in_bytes` and `memory.max_usage_in_bytes` match up fine though, presumably because swap isn't used.

**Comment by nuwang** [comment]
Closed by: #12792
